src/Stratis.Bitcoin.IntegrationTests/EnvironmentMockUpHelpers/MockedInitialBlockDownloadState.cs

```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Configuration;
+using Stratis.Bitcoin.Features.Consensus;
+
+namespace Stratis.Bitcoin.IntegrationTests.EnvironmentMockUpHelpers
+{
+    public class MockedInitialBlockDownloadState : InitialBlockDownloadState
```

Comment:
@noescape00 I'm not too sure that mocks should inherit from the actual implementation like this? I'm used to a mocks implementing an interface and defining (custom) logic in contract members and/or methods. It'll mean possibly duplicating the logic in IsInitialBlockDownload() but is all that logic necessary for the unit tests when in fact we are mocking the IBD provider?

Also, when we call NotInIBD:

` (this.FullNode.NodeService<IInitialBlockDownloadState>() as MockedInitialBlockDownloadState).SetIsInitialBlockDownload(false, DateTime.UtcNow.AddMinutes(5));`

We don't have to cast.

**Also**, I would rename to `MockedInitialBlockDownloadState` to `InitialBlockDownloadStateMock`.

Reply:
>I'm not too sure that mocks should inherit from the actual implementation like this?

Yes, that was the intention. In some tests we actually want a real IBD state, in some we would like to ignore it. So unless mocked IBD state is set this one will act exactly like real environment IBD state provider. 

> It'll mean possibly duplicating the logic in IsInitialBlockDownload() 

You mean implementing just the interface and copy paste the rest? I though about this but if we do it like that a scenario when we change full node IBD state provider and forget about the test mock is possible and we would like to avoid that. 

>We don't have to cast.

'SetIsInitialBlockDownload' is not a part of the interface so we kind of have to. Or did I miss something?

Reply:
>Also, I would rename to MockedInitialBlockDownloadState to InitialBlockDownloadStateMock.

Good idea, will do.

Reply:
>You mean implementing just the interface and copy paste the implementation? 

Correct.

> I though about this but if we do it like that a scenario when we change full node IBD state provider and forget about the test mock is possible so we would like to avoid that.

You will still register the mock in feature registration for unit tests, as you did now, it's just that the mock will implement the interface and not inherit from the actual implementation. 

I'm not too much of a fan of making methods virtual for the sake of unit tests. Because you are implying that logic in there can be overridden in production code. That's why implementing the interface is better design for creating a mock test class. It feels funny extending production code in test code, that's my feeling on this. Yes we are duplicating logic but if we ever delete IBD provider, the actual provider will remain unchanged.

> 'SetIsInitialBlockDownload' is not a part of the interface so we kind of have to. Or did I miss something?

Sorry you are right :)

@noescape00 I'm just thinking out load here, so am open to discussion. What do you think?

Reply:
>I'm not too much of a fan of making methods virtual for the sake of unit tests. 

Hmm, makes sense. 

I think we can do it even better: avoid copy pasting the implementation code and avoid making the method virtual for the production code by just making an inner IBD state provider inside the mock.

Reply:
Fixed. 
@fassadlr let me know if you're happy with the new approach